aggregator/aggregator.go

```diff
@@ -43,275 +40,238 @@ func NewAggregator(
+		return
+	}
+
+	if !isProfitable {
+		log.Infof("Batch %d is not profitable, matic collateral %d", batchToVerify.BatchNumber, big.NewInt(0))
+		return
+	}
+
+	log.Infof("sending zki + batch to the prover, batchNumber: %d", batchToVerify.BatchNumber)
+	inputProver, err := a.buildInputProver(ctx, batchToVerify)
+	if err != nil {
+		log.Warnf("failed to build input prover, err: %v", err)
+		return
+	}
+
+	genProofID, err := a.ProverClient.GetGenProofID(ctx, inputProver)
+	if err != nil {
+		log.Warnf("failed to get gen proof id, err: %v", err)
+		return
+	}
+
+	resGetProof, err := a.ProverClient.GetResGetProof(ctx, genProofID, batchToVerify.BatchNumber)
+	if err != nil {
+		log.Warnf("failed to get proof from prover, err: %v", err)
+		return
+	}
+	a.compareInputHashes(inputProver, resGetProof)
+
+	log.Infof("sending verified proof to the ethereum smart contract, batchNumber", batchToVerify.BatchNumber)
+	err = a.EthTxManager.VerifyBatch(batchToVerify.BatchNumber, resGetProof)
+	if err != nil {
+		log.Warnf("failed to send request to consolidate batch to ethereum, batch number: %d, err: %v",
+			batchToVerify.BatchNumber, err)
+		return
+	}
+	a.batchesSent[batchToVerify.BatchNumber] = true
+	log.Infof("proof for the batch was send, batchNumber: %d", batchToVerify.BatchNumber)
+}
+
+func (a *Aggregator) isSynced(ctx context.Context) bool {
+	lastVerifiedBatch, err := a.State.GetLastVerifiedBatch(ctx, nil)
+	if err != nil && err != state.ErrNotFound {
+		log.Warnf("failed to get last consolidated batch, err: %v", err)
+		return false
+	}
+	if lastVerifiedBatch != nil {
+		a.lastVerifiedBatchNum = lastVerifiedBatch.BatchNumber
+	}
+	lastVerifiedEthBatchNum, err := a.Ethman.GetLatestVerifiedBatchNum()
+	if err != nil {
+		log.Warnf("failed to get last eth batch, err: %v", err)
+		return false
+	}
+	if a.lastVerifiedBatchNum < lastVerifiedEthBatchNum {
+		log.Infof("waiting for the state to be synced, lastVerifiedBatchNum: %d, lastVerifiedEthBatchNum: %d",
+			a.lastVerifiedBatchNum, lastVerifiedEthBatchNum)
+		return false
+	}
+	return true
+}
+
+func (a *Aggregator) getBatchToVerify(ctx context.Context) (*state.Batch, error) {
+	delete(a.batchesSent, a.lastVerifiedBatchNum)
+
+	batchToVerify, err := a.State.GetVirtualBatchByNumber(ctx, a.lastVerifiedBatchNum+1, nil)
+
+	if err != nil {
+		if errors.Is(err, state.ErrNotFound) {
+			log.Infof("there are no batches to consolidate")
+			return nil, err
 		}
+		log.Warnf("failed to get batch to consolidate, err: %v", err)
+		return nil, err
+	}
+
+	if a.batchesSent[batchToVerify.BatchNumber] {
+		log.Infof("batch with number %d was already sent, but not yet consolidated by synchronizer",
+			batchToVerify.BatchNumber)
+		return nil, nil
+	}
+
+	return batchToVerify, nil
+}
+
+func (a *Aggregator) buildInputProver(ctx context.Context, batchToVerify *state.Batch) (*pb.InputProver, error) {
+	oldStateRoot, err := a.State.GetStateRootByBatchNumber(ctx, a.lastVerifiedBatchNum, nil)
```

Comment:
```suggestion
	oldStateRoot, err := a.State.GetStateRootByBatchNumber(ctx, batchToVerify.BatchNumber -1, nil)
```

Reply:
Actually, it would be better if you get the full batch, so you avoid extra DB calls later on, for instance calling `GetLocalExitRootByBatchNumber`